File: flutter/lib/views/postsViewModel.dart
import 'package:SocialNetwork/domain/post/reaction.dart';
import 'package:SocialNetwork/domain/post/usecases/reactToPostUseCase.dart';

import '../data/database.dart';
import '../data/post/postDao.dart';
import '../data/post/postRepository.dart';
import '../data/post/postService.dart';
import '../domain/post/postRepostiory.dart';
import '../domain/post/usecases/createPostUseCase.dart';
import '../domain/post/usecases/getAllPostsUseCase.dart';
import 'package:flutter/material.dart';
import '../domain/post/post.dart';
import '../domain/user/user.dart';
import 'bindingTextField.dart';

class PostsViewModel extends ChangeNotifier {
  Binding<String> post = Binding.initialValue("");
  List<Post> posts = List();
  String error;

  PostRepository postRepository;

  PostsViewModel() {
    this.postRepository = PostRepositoryImpl(PostDao(AppDatabase()), PostService());
  }

  void loadData() async {
    try {
      this.posts = await GetAllPostsUseCase(postRepository).run();
    } catch (error) {
      this.error = error.toString();
    }
    notifyListeners();
  }

  void createPost() async {
    Post newPost = Post(user: User.currentUser, text: post.value, date: DateTime.now().millisecondsSinceEpoch);
    try {
      Post storedPost = await CreatePostUseCase(postRepository).run(newPost);
      posts.add(storedPost);
    } catch (error) {
      this.error = error.toString();
    }
    notifyListeners();
  }

  void reactToPost(Post post, ReactionType reactionType) async {
    Reaction reaction = Reaction(post, reactionType, User.currentUser, DateTime.now().millisecondsSinceEpoch);
    try {
      Post response = await ReactToPostUseCase(postRepository).run(reaction);
      int indexOf = posts.indexWhere((post) => response.id == post.id);
      posts.replaceRange(indexOf, indexOf, [response]);
    } catch (error) {
      this.error = error.toString();
    }
    notifyListeners();
  }
}

File: web/src/views/Posts/PostsViewModel.ts
import ViewModel from "../ViewModel";
import Post from "../../domain/post/Post";
import PostRepositoryImpl from "../../data/PostRepositoryImpl";
import PostService from "../../data/PostService";
import PostRepository from "../../domain/post/PostRepository";
import GetAllPostsUseCase from "../../domain/user/usecases/GetAllPostsUseCase";
import CreatePostUseCase from "../../domain/user/usecases/CreatePostUseCase";
import GetLoggedUserUseCase from "../../domain/post/usecases/GetLoggedUserUseCase";
import UserRepositoryImpl from "../../data/UserRepositoryImpl";
import UserService from "../../data/UserService";
import Reaction from "../../domain/post/Reaction";
import ReactToPostUseCase from "../../domain/user/usecases/ReactToPostUseCase";

export type State = {
    loading: boolean;
    posts: Array<Post> | null;
    post: string;
}

export class PostsViewModel extends ViewModel {
    postRepository: PostRepository;

    constructor() {
        super();
        this.postRepository = new PostRepositoryImpl(new PostService());

        this.state = {
            loading: false,
            posts: null,
            post: '',
        }
    }

    async loadPosts() {
        let posts = await new GetAllPostsUseCase(this.postRepository).run();
        this.setState({
            posts,
        });
    }

    setPost = (post: string | null) => {
        this.setState({
            post,
        });
    }

    async createPost() {
        console.log(this.state);
        const user = await new GetLoggedUserUseCase(new UserRepositoryImpl(new UserService())).run();
        const post = new Post(
            null,
            user!,
            this.state.post,
            new Date().valueOf(),
            null
        );
        const result = await new CreatePostUseCase(this.postRepository).run(post);
        const array = this.state.posts || [];
        array.push(result);
        this.setState({
            posts: array,
        });
    }

    async reactToPost(post: Post, reactionType: string) {
        const user = await new GetLoggedUserUseCase(new UserRepositoryImpl(new UserService())).run();
        const reaction = new Reaction(
            post.id!,
            reactionType,
            user!,
            new Date().valueOf()
        )
        const result = await new ReactToPostUseCase(this.postRepository).run(reaction);
        const updatedPost = Post.fromJson(result);

        const postIndex = this.state.posts!.findIndex((existingPost: Post) => {
            return existingPost.id === post.id;
        });
        this.updatePostAtIndex(postIndex, updatedPost);
    }

    updatePostAtIndex(postIndex: number, updatedPost: Post) {
        const array = this.state.posts!;
        array[postIndex] = updatedPost;
        this.setState({
            posts: array,
        });
    }
}